src/DynamicRupture/FrictionLaws/GpuImpl/GpuBaseFrictionLaw.cpp

```diff
@@ -4,108 +4,53 @@
-    seissol::dr::friction_law::FrictionSolver::savePeakSlipRateOutput,                             \
-    seissol::dr::friction_law::FrictionSolver::saveRuptureFrontOutput)
-
-// clang-format off
 namespace seissol::dr::friction_law::gpu {
 GpuBaseFrictionLaw::GpuBaseFrictionLaw(dr::DRParameters& drParameters)
     : FrictionSolver(drParameters) {
   checkOffloading();
 }
 
 GpuBaseFrictionLaw::~GpuBaseFrictionLaw() {
-  //#pragma omp target exit data map(release:faultStresses[0:maxClusterSize])
-  delete[] faultStresses;
-  delete[] tractionResults;
+  omp_target_free(faultStresses, diviceId);
+  omp_target_free(tractionResults, diviceId);
+  omp_target_free(stateVariableBuffer, diviceId);
+  omp_target_free(strengthBuffer, diviceId);
+  omp_target_free(resampleMatrix, diviceId);
 }
 
-void GpuBaseFrictionLaw::evaluate(seissol::initializers::Layer& layerData,
-                                  seissol::initializers::DynamicRupture* dynRup,
-                                  real fullUpdateTime,
-                                  double timeWeights[CONVERGENCE_ORDER]) {
-  FrictionSolver::copyLtsTreeToLocal(layerData, dynRup, fullUpdateTime);
-  this->copySpecificLtsDataTreeToLocal(layerData, dynRup, fullUpdateTime);
-
-  #pragma omp target data map(to : this)
-  {
-    auto layerSize = layerData.getNumberOfCells();
-
-    #pragma omp target teams loop \
-    map(from: faultStresses [0:layerSize]) \
-    is_device_ptr(qInterpolatedPlus, qInterpolatedMinus, impAndEta) \
-    device(diviceId)
-    for (unsigned ltsFace = 0; ltsFace < layerSize; ++ltsFace) {
-      precomputeStressFromQInterpolated(faultStresses[ltsFace], ltsFace);
-    }
-
-    // loop over all dynamic rupture faces, in this LTS layer
-    #pragma omp parallel for schedule(static)
-    for (unsigned ltsFace = 0; ltsFace < layerSize; ++ltsFace) {
-
-      // define some temporary variables
-      std::array<real, misc::numPaddedPoints> stateVariableBuffer{0};
-      std::array<real, misc::numPaddedPoints> strengthBuffer{0};
-
-      this->preHook(stateVariableBuffer, ltsFace);
-
-      // loop over sub time steps (i.e. quadrature points in time)
-      for (unsigned timeIndex = 0; timeIndex < CONVERGENCE_ORDER; timeIndex++) {
-        this->updateFrictionAndSlip(faultStresses[ltsFace],
-                                    tractionResults[ltsFace],
-                                    stateVariableBuffer,
-                                    strengthBuffer,
-                                    ltsFace,
-                                    timeIndex);
-      }
+void GpuBaseFrictionLaw::allocateAuxiliaryMemory(seissol::initializers::LTSTree* drTree,
+                                                 seissol::initializers::DynamicRupture* drDescr,
+                                                 int currDiviceId) {
+  for (auto it = drTree->beginLeaf(seissol::initializers::LayerMask(Ghost));
```

Comment:
I moved it to `LTSTree class`

```
  size_t getMaxClusterSize(LayerMask mask) {
    size_t maxClusterSize{0};
    for (auto it = beginLeaf(mask); it != endLeaf(); ++it) {
      size_t currClusterSize = static_cast<size_t>(it->getNumberOfCells());
      maxClusterSize = std::max(currClusterSize, maxClusterSize);
    }
    return maxClusterSize;
  }
```

because it logically fits there.


```
void seissol::initializers::MemoryManager::initFrictionData(seissol::Interoperability *interoperability) {
  if (m_dynRupParameter.isDynamicRuptureEnabled) {

    m_DRInitializer->initializeFault(m_dynRup.get(), &m_dynRupTree, interoperability);

#ifdef ACL_DEVICE_OFFLOAD
    if (auto* impl = dynamic_cast<dr::friction_law::gpu::GpuBaseFrictionLaw*>(m_FrictionLaw.get())) {
      device::DeviceInstance& device = device::DeviceInstance::getInstance();
      impl->setDeviceId(device.api->getDeviceId());

      LayerMask mask = seissol::initializers::LayerMask(Ghost);
      auto maxSize = m_dynRupTree.getMaxClusterSize(mask);
      impl->setMaxClusterSize(maxSize);

      impl->initializeAuxiliaryMemory();
    }
#endif // ACL_DEVICE_OFFLOAD

    interoperability->initializeFaultOutput();
  }
}
```